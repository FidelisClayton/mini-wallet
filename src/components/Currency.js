import React from 'react'
import PropTypes from 'prop-types'
import { css } from 'emotion'

import Button from './Button'
import ArrowIcon from './ArrowIcon'

import * as colors from '../helpers/colors'

const styles = css({
  display: 'flex',
  marginTop: '10px',
  borderRadius: '5px',
  backgroundColor: colors.white,
  border: colors.cardBorder,

  '.currency': {
    '&__data': {
      display: 'flex',
      flexDirection: 'column',
      flexGrow: 1,

      '&--right': {
        textAlign: 'right'
      }
    },

    '&__info': {
      display: 'flex',
      paddingTop: '15px',
      paddingBottom: '15px',
      paddingLeft: '15px',
      paddingRight: '15px',
      flex: 1,
      width: '100%',
    },

    '&__actions': {
      display: 'flex'
    },

    '&__action-button': {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      justifyContent: 'center',
      height: '100%',
      borderRadius: 0,
      width: '70px',
      fontSize: '0.7rem',
      fontWeight: 'normal',
      textTransform: 'uppercase',
      border: 0,
      outline: 'none',

      '&--sell': {
        background: colors.darkBlue
      }
    },

    '&__action-icon': {
      marginBottom: '5px',
      width: '15px',
      height: '15px',

      '&--sell': {
        transform: 'rotate(180deg)'
      }
    },

    '&__coin, &__total': {
      fontWeight: 'bold',
      color: colors.darkBlue
    },

    '&__price, &__amount': {
      fontWeight: 'bold',
      fontSize: '0.8rem',
      color: colors.lightBlue
    }
  }
})

const Currency = ({
  name,
  price,
  total,
  amount,
  currency,
  exchangeable,
  onSell,
  ...props
}) => {
  return (
    <div
      className={`currency ${styles}`}
      {...props}
    >
      <div className="currency__info">
        <div className="currency__data">
          <label className="currency__label currency__coin">{ name }</label>

          { exchangeable && price && (
            <label className="currency__label currency__price">R$ { price }</label>
          )}
        </div>

        <div className="currency__data currency__data--right">
          <label className="currency__label currency__total">R$ { total }</label>
          { amount !== null && (
            <label className="currency__label currency__amount">{ currency } { amount }</label>
          )}
        </div>
      </div>

      { exchangeable && (
        <div className="currency__actions">
          <Button className="currency__action-button">
            <ArrowIcon className="currency__action-icon" />
            Comprar
          </Button>
          <Button
            className="currency__action-button currency__action-button--sell"
            onClick={onSell}
          >
            <ArrowIcon className="currency__action-icon currency__action-icon--sell" />
            Vender
          </Button>
        </div>
      )}
    </div>
  )
}

Currency.defaultProps = {
  exchangeable: true
}

Currency.propTypes = {
  name: PropTypes.string.isRequired,
  price: PropTypes.number,
  total: PropTypes.number,
  amount: PropTypes.amount,
  styles: PropTypes.string
}

export default Currency
