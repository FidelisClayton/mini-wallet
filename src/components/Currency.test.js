import React from 'react'
import { shallow } from 'enzyme'

import Currency from './Currency'

const bitcoin = {
  name: 'Bitcoin',
  currency: 'BTC',
  price: 29070,
  total: 2 * 29070,
  amount: 2,
  exchangeable: true,
  onSell: jest.fn(),
  onBuy: jest.fn()
}

const real = {
  name: 'Real',
  amount: 100000,
  currency: 'BRL',
  exchangeable: false
}

const brita = {
  name: 'Brita',
  currency: 'BRT',
  price: 3.76,
  total: 2 * 3.76,
  amount: null,
  exchangeable: true,
  onSell: jest.fn(),
  onBuy: jest.fn()
}

describe('Component: <Currency />', () => {
  it('should properly render', () => {
    const wrapper = shallow(
      <Currency {...bitcoin} />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should not render action buttons for a non exchangeable coin', () => {
    const wrapper = shallow(
      <Currency {...real} />
    )

    expect(wrapper.find('.currency__actions')).toHaveLength(0)
  })

  it('should not display price for a non exchangeable coin', () => {
    const wrapper = shallow(
      <Currency {...real} />
    )

    expect(wrapper.find('.currency__price')).toHaveLength(0)
  })

  it('should not render amount if the amount value is not provided', () => {
    const wrapper = shallow(
      <Currency {...brita} />
    )

    expect(wrapper.find('.currency__amount')).toHaveLength(0)
  })

  it('should handle buy click', () => {
    const onBuy = jest.fn()

    const wrapper = shallow(
      <Currency
        {...bitcoin}
        onBuy={onBuy}
      />
    )

    wrapper.find('.currency__buy').simulate('click')

    expect(onBuy.mock.calls).toHaveLength(1)
  })

  it('should handle sell click', () => {
    const onSell = jest.fn()

    const wrapper = shallow(
      <Currency
        {...bitcoin}
        onSell={onSell}
      />
    )

    wrapper.find('.currency__sell').simulate('click')

    expect(onSell.mock.calls).toHaveLength(1)
  })
})
