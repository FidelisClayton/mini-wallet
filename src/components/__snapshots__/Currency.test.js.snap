// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Currency /> should properly render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Currency
    amount={2}
    currency="BTC"
    exchangeable={true}
    name="Bitcoin"
    onBuy={[Function]}
    onSell={[Function]}
    price={29070}
    total={58140}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="currency__info"
>
          <div
                    className="currency__data"
          >
                    <label
                              className="currency__label currency__coin"
                    >
                              Bitcoin
                    </label>
          </div>
          <div
                    className="currency__data currency__data--right"
          >
                    <label
                              className="currency__label currency__total"
                    >
                              R$ 
                              58.140,00
                    </label>
                    <label
                              className="currency__label currency__amount"
                    >
                              BTC
                               
                              2,00000000
                    </label>
          </div>
</div>,
        <div
          className="currency__actions"
>
          <Button
                    className="currency__action-button currency__buy"
                    onClick={[Function]}
          >
                    <DownArrow
                              className="currency__action-icon"
                    />
                    Comprar
          </Button>
          <Button
                    className="currency__action-button currency__sell currency__action-button--sell"
                    onClick={[Function]}
          >
                    <DownArrow
                              className="currency__action-icon currency__action-icon--sell"
                    />
                    Vender
          </Button>
</div>,
      ],
      "className": "currency css-1aypruw",
      "price": 29070,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="currency__data"
>
              <label
                            className="currency__label currency__coin"
              >
                            Bitcoin
              </label>
</div>,
            <div
              className="currency__data currency__data--right"
>
              <label
                            className="currency__label currency__total"
              >
                            R$ 
                            58.140,00
              </label>
              <label
                            className="currency__label currency__amount"
              >
                            BTC
                             
                            2,00000000
              </label>
</div>,
          ],
          "className": "currency__info",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  className="currency__label currency__coin"
>
                  Bitcoin
</label>,
                undefined,
              ],
              "className": "currency__data",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Bitcoin",
                  "className": "currency__label currency__coin",
                },
                "ref": null,
                "rendered": "Bitcoin",
                "type": "label",
              },
              undefined,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  className="currency__label currency__total"
>
                  R$ 
                  58.140,00
</label>,
                <label
                  className="currency__label currency__amount"
>
                  BTC
                   
                  2,00000000
</label>,
              ],
              "className": "currency__data currency__data--right",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "R$ ",
                    "58.140,00",
                  ],
                  "className": "currency__label currency__total",
                },
                "ref": null,
                "rendered": Array [
                  "R$ ",
                  "58.140,00",
                ],
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "BTC",
                    " ",
                    "2,00000000",
                  ],
                  "className": "currency__label currency__amount",
                },
                "ref": null,
                "rendered": Array [
                  "BTC",
                  " ",
                  "2,00000000",
                ],
                "type": "label",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Button
              className="currency__action-button currency__buy"
              onClick={[Function]}
>
              <DownArrow
                            className="currency__action-icon"
              />
              Comprar
</Button>,
            <Button
              className="currency__action-button currency__sell currency__action-button--sell"
              onClick={[Function]}
>
              <DownArrow
                            className="currency__action-icon currency__action-icon--sell"
              />
              Vender
</Button>,
          ],
          "className": "currency__actions",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <DownArrow
                  className="currency__action-icon"
/>,
                "Comprar",
              ],
              "className": "currency__action-button currency__buy",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "currency__action-icon",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "Comprar",
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <DownArrow
                  className="currency__action-icon currency__action-icon--sell"
/>,
                "Vender",
              ],
              "className": "currency__action-button currency__sell currency__action-button--sell",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "currency__action-icon currency__action-icon--sell",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "Vender",
            ],
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="currency__info"
>
            <div
                        className="currency__data"
            >
                        <label
                                    className="currency__label currency__coin"
                        >
                                    Bitcoin
                        </label>
            </div>
            <div
                        className="currency__data currency__data--right"
            >
                        <label
                                    className="currency__label currency__total"
                        >
                                    R$ 
                                    58.140,00
                        </label>
                        <label
                                    className="currency__label currency__amount"
                        >
                                    BTC
                                     
                                    2,00000000
                        </label>
            </div>
</div>,
          <div
            className="currency__actions"
>
            <Button
                        className="currency__action-button currency__buy"
                        onClick={[Function]}
            >
                        <DownArrow
                                    className="currency__action-icon"
                        />
                        Comprar
            </Button>
            <Button
                        className="currency__action-button currency__sell currency__action-button--sell"
                        onClick={[Function]}
            >
                        <DownArrow
                                    className="currency__action-icon currency__action-icon--sell"
                        />
                        Vender
            </Button>
</div>,
        ],
        "className": "currency css-1aypruw",
        "price": 29070,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="currency__data"
>
                <label
                                className="currency__label currency__coin"
                >
                                Bitcoin
                </label>
</div>,
              <div
                className="currency__data currency__data--right"
>
                <label
                                className="currency__label currency__total"
                >
                                R$ 
                                58.140,00
                </label>
                <label
                                className="currency__label currency__amount"
                >
                                BTC
                                 
                                2,00000000
                </label>
</div>,
            ],
            "className": "currency__info",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    className="currency__label currency__coin"
>
                    Bitcoin
</label>,
                  undefined,
                ],
                "className": "currency__data",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Bitcoin",
                    "className": "currency__label currency__coin",
                  },
                  "ref": null,
                  "rendered": "Bitcoin",
                  "type": "label",
                },
                undefined,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <label
                    className="currency__label currency__total"
>
                    R$ 
                    58.140,00
</label>,
                  <label
                    className="currency__label currency__amount"
>
                    BTC
                     
                    2,00000000
</label>,
                ],
                "className": "currency__data currency__data--right",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "R$ ",
                      "58.140,00",
                    ],
                    "className": "currency__label currency__total",
                  },
                  "ref": null,
                  "rendered": Array [
                    "R$ ",
                    "58.140,00",
                  ],
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "BTC",
                      " ",
                      "2,00000000",
                    ],
                    "className": "currency__label currency__amount",
                  },
                  "ref": null,
                  "rendered": Array [
                    "BTC",
                    " ",
                    "2,00000000",
                  ],
                  "type": "label",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Button
                className="currency__action-button currency__buy"
                onClick={[Function]}
>
                <DownArrow
                                className="currency__action-icon"
                />
                Comprar
</Button>,
              <Button
                className="currency__action-button currency__sell currency__action-button--sell"
                onClick={[Function]}
>
                <DownArrow
                                className="currency__action-icon currency__action-icon--sell"
                />
                Vender
</Button>,
            ],
            "className": "currency__actions",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <DownArrow
                    className="currency__action-icon"
/>,
                  "Comprar",
                ],
                "className": "currency__action-button currency__buy",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "currency__action-icon",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "Comprar",
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <DownArrow
                    className="currency__action-icon currency__action-icon--sell"
/>,
                  "Vender",
                ],
                "className": "currency__action-button currency__sell currency__action-button--sell",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "currency__action-icon currency__action-icon--sell",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "Vender",
              ],
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
